### PR TITLE
Improve support for out-of-tree installations

### PR DESCRIPTION
**Pull request by nmnobre**
Hi,

The changes here have the objective of allowing a classical build (i.e. one via the provided `makefile`) to install the required files on a system directory (via `PREFIX=`) with suitable permissions for use by everyone in that system.
In addition, a user (or a module file for instance) should then be able to set `MFEM_INSTALL_DIR` allowing the use of the provided `makefile`s to build examples, miniapps and tests. Note the idea to use `or` and `wildcard` comes from an existing `makefile`:

https://github.com/mfem/mfem/blob/12ed616e47ea47c6d36a4c40290cf41510544914/tests/par-mesh-format/makefile#L16-L17

Cheers,
-Nuno

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4252](https://github.com/mfem/mfem/pull/4252) | @nmnobre | @tzanio | @justinlaughlin + @tzanio + @v-dobrev | 5/11/24 | 7/16/24 | 7/19/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
Hi @nmnobre, if this is ready, can you mark it with the `ready-for-review` label please?

(See also the [CONTRIBUTING](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#quick-summary) document.)

**Comment by tzanio** [comment]
> > Hi @nmnobre, if this is ready, can you mark it with the `ready-for-review` label please?
> 
> I can't, I don't think I have triage access on this repo.

Can you please try again?

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: makefile
Comment (justinlaughlin) [inline makefile:1]: This _seems_ okay to me but there are a lot of changed permissions so I'd like to get a second pair of eyes - @v-dobrev do you have any input?

Main things I'm noticing is 
 - `mkdir -p $(DIR)` changed to `$(INSTALL) -m 755 -d $(DIR)`. `mkdir` defaults to 755 so this seems okay to me?
 - Granting read privileges to all users (640 -> 644)
 - Lines 632 and 635 change the privileges from 640 -> 755, is this necessary?

Comment (nmnobre) [reply makefile:1]: Hi Justin, thanks for taking the time to review this PR :)

> `mkdir -p $(DIR)` changed to `$(INSTALL) -m 755 -d $(DIR)`

Indeed, this should be equivalent on most (all?) platforms. I just thought... since we're using `$(INSTALL)` for files why not follow the same conventional industry practices for directories too?

> Granting read privileges to all users (640 -> 644)

I tried to follow what other people are doing. The llvm guys install `.a` libs with 644, `.so` libs with 755 and `.h` files with 644.

> Lines 632 and 635 change the privileges from 640 -> 755, is this necessary?

The people at Nvidia install `.mk` files with 755, but Intel uses 644. I think I used the former just because I happened to come across the Nvidia makefiles first (on closer inspection, Nvidia install way too many things with 755... including header files??!!), but I'm happy to change to 644, I agree it'd make more sense.

Comment (justinlaughlin) [reply makefile:1]: Hi Numo, sorry for the delayed reply.

Would you be able to add a variable to toggle back and forth between these new permissions and the old ones (with the old settings set as the default)? The main concern is we don't want users to install using these new permissions without intending to - being mindful that requirements/risk-averseness at a government lab might be different from those at Nvidia or Intel :). Otherwise, this is a great PR - thank you for your contribution!

Comment (nmnobre) [reply makefile:1]: Hi Justin (@justinlaughlin),

I think I complicated my argument a bit too much.
It's this simple: these are the *exact* same permissions you get if you use the cmake build system, no more, no less. I don't understand what risks this poses either? The software is open-source, there's nothing the users would be able to read they wouldn't be able to read by coming to the source on GitHub?
The permission changes suggested here simply make both build systems consistent from the user's perspective, and I know I'm not the only one who likes consistency around here. :)

File: makefile [hunk 1]
```diff
@@ -587,51 +587,52 @@ distclean: clean config/clean doc/clean
...
 # install config.mk in $(PREFIX_SHARE)
-	mkdir -p $(PREFIX_SHARE)
+	$(INSTALL) -m 755 -d $(PREFIX_SHARE)
 	$(MAKE) -C $(BLD)config config-mk CONFIG_MK=config-install.mk
-	$(INSTALL) -m 640 $(BLD)config/config-install.mk $(PREFIX_SHARE)/config.mk
+	$(INSTALL) -m 644 $(BLD)config/config-install.mk $(PREFIX_SHARE)/config.mk
 	rm -f $(BLD)config/config-install.mk
 # install test.mk in $(PREFIX_SHARE)
-	$(INSTALL) -m 640 $(SRC)config/test.mk $(PREFIX_SHARE)/test.mk
+	$(INSTALL) -m 644 $(SRC)config/test.mk $(PREFIX_SHARE)/test.mk
```
Comment (v-dobrev) [inline makefile:635]: Hi @nmnobre,

Here's what I suggest we use with regard to install permissions -- it is very similar to what you have; compared to the current `master`, it uses the same permission for "others" as for "group"; it also allows easy switch to other permissions by setting the make variables `INSTALL_DEF_PERM`, `INSTALL_BIN_PERM`, and `INSTALL_DIR_MASK` on the command line or in `user.mk`

```suggestion
# Install permissions for everything except directories and binaries:
INSTALL_DEF_PERM ?= 644
# Install permissions for binaries:
INSTALL_BIN_PERM ?= 755
# Install umask for creating directories (and symlinks on MacOS/BSD).
# We use 'umask' because 'mkdir -p' (and 'install -d') do not use the mode
# specified with the '-m' flag when creating non-existent parent directories.
INSTALL_DIR_MASK ?= 022

INSTALL_SHARED_LIB = $(MFEM_CXX) $(MFEM_LINK_FLAGS) $(INSTALL_SOFLAGS)\
   $(OBJECT_FILES) $(EXT_LIBS) -o $(PREFIX_LIB)/libmfem.$(SO_VER) && \
   cd $(PREFIX_LIB) && chmod $(INSTALL_BIN_PERM) libmfem.$(SO_VER) && \
   ( umask $(INSTALL_DIR_MASK) && ln -sf libmfem.$(SO_VER) libmfem.$(SO_EXT) )

# Shortcuts, not to be modified on command line.
# WARNING: $(MKINSTALLDIR) changes the umask for commands following it as part
#          of the same shell expression unless it is placed inside '()' to be
#          executed in a sub-shell.
override MKINSTALLDIR = umask $(INSTALL_DIR_MASK) && mkdir -p
override INSTALLDEF   = $(INSTALL) -m $(INSTALL_DEF_PERM)

install: $(if $(static),$(BLD)libmfem.a) $(if $(shared),$(BLD)libmfem.$(SO_EXT))
	$(MKINSTALLDIR) $(PREFIX_LIB)
# install static and/or shared library
	$(if $(static),$(INSTALLDEF) $(BLD)libmfem.a $(PREFIX_LIB))
	$(if $(shared),$(INSTALL_SHARED_LIB))
# install top level includes
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem
	$(INSTALLDEF) $(SRC)mfem.hpp $(SRC)mfem-performance.hpp \
	   $(PREFIX_INC)/mfem
	for hdr in mfem.hpp mfem-performance.hpp; do \
	   printf '// Auto-generated file.\n#include "mfem/'$$hdr'"\n' \
	      > $(PREFIX_INC)/$$hdr && \
	   chmod $(INSTALL_DEF_PERM) $(PREFIX_INC)/$$hdr; done
# install config include
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/config
	$(INSTALLDEF) $(BLD)config/_config.hpp $(PREFIX_INC)/mfem/config
	$(INSTALLDEF) $(SRC)config/config.hpp $(PREFIX_INC)/mfem/config
	$(INSTALLDEF) $(SRC)config/tconfig.hpp $(PREFIX_INC)/mfem/config
# install remaining includes in each subdirectory
	for dir in $(DIRS); do \
	   ( $(MKINSTALLDIR) $(PREFIX_INC)/mfem/$$dir ) && \
	   $(INSTALLDEF) $(SRC)$$dir/*.hpp $(PREFIX_INC)/mfem/$$dir; \
	done
# install *.okl files
	for dir in $(OKL_DIRS); do \
	   ( $(MKINSTALLDIR) $(PREFIX_INC)/mfem/$$dir ) && \
	   $(INSTALLDEF) $(SRC)$$dir/*.okl $(PREFIX_INC)/mfem/$$dir; \
	done
# install libCEED q-function headers
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/mass/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/convection/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/diffusion/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/nlconvection/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
# install config.mk in $(PREFIX_SHARE)
	$(MKINSTALLDIR) $(PREFIX_SHARE)
	$(MAKE) -C $(BLD)config config-mk CONFIG_MK=config-install.mk
	$(INSTALLDEF) $(BLD)config/config-install.mk $(PREFIX_SHARE)/config.mk
	rm -f $(BLD)config/config-install.mk
# install test.mk in $(PREFIX_SHARE)
	$(INSTALLDEF) $(SRC)config/test.mk $(PREFIX_SHARE)
```

File: miniapps/dpg/makefile [hunk 2]
```diff
@@ -12,6 +12,7 @@
 # Use the MFEM build directory
 MFEM_DIR ?= ../..
 MFEM_BUILD_DIR ?= ../..
+MFEM_INSTALL_DIR ?= ../../mfem
 SRC = $(if $(MFEM_DIR:../..=),$(MFEM_DIR)/miniapps/dpg/,)
 CONFIG_MK = $(MFEM_BUILD_DIR)/config/config.mk
```
Comment (v-dobrev) [inline miniapps/dpg/makefile:17]: Update the definition of `CONFIG_MK` here as well.

Comment (nmnobre) [reply miniapps/dpg/makefile]: My bad. Thanks for spotting this!

File: makefile (hunk 1 above)
Comment (nmnobre) [reply makefile]: I like your idea, so here's what I suggest (and just committed):

- We roll back to `mkdir -p` to create the necessary directories. This sets the same permissions for parent directories and honours the user's umask. This was never a problem, so perhaps I should've never touched it. My bad.
- We use `cp` to copy libraries, headers and makefiles to the new directories. This will always subtract the user's umask from the source files permissions. Then, we just do `chmod +x` for the shared library file.

I think everybody will be happy with this solution:
- Typical users, like me, who probably never touched their umask, so it's still at 022, will get 755 for all directories and 644 for most files (except the shared object). This is what I expect will happen most often.
- People at gov labs might've different umask settings (maybe 026), either set by themselves or by some system administrator, and these will still be honoured with the suggested solution. Most files will then be installed with 640 as before, and directories with 751.
- It avoids the extra makefile variables, the user can simply tweak their umask to get the desired permissions.
 
How does that sound?

Comment (v-dobrev) [reply makefile]: I don't think we want to switch from `$(INSTALL)` to `cp` -- I know that users use the `INSTALL` variable to adjust things; and we don't want to create confusion by setting `INSTALL` to `cp`. Also, having the explicit variables for permissions makes it more obvious for users how to adjust those without having to figure out how they can have some control by modifying their `umask`.

Also, I have a tweak in the shared library installation command that I want to include (it affects MacOS).

Basically, I'll re-iterate my suggestion to apply the changes I posted above.

Comment (nmnobre) [reply makefile]: Let me know what you think of this. Instead of setting `INSTALL_DIR_MASK`, the user may set `INSTALL_DIR_PERM` which I think might be less confusing and follows the same format of the other two new variables. Otherwise, it's as you suggested. With your permission, I'd like to squash some commits and force push.

Comment (v-dobrev) [reply makefile]: > Let me know what you think of this.

Looks good to me. Thanks!

> Instead of setting `INSTALL_DIR_MASK`, the user may set `INSTALL_DIR_PERM` which I think might be less confusing and follows the same format of the other two new variables.

Yes, this is a good change.

> Otherwise, it's as you suggested. If you're happy with this, and with your permission, I'd like to squash some commits and force push.

We can also just squash-merge the PR. If, however, you prefer to have a sequence of commits from the PR, I don't mind if you want to force-push. Just make sure the editor, @tzanio, and the other reviewers, @justinlaughlin, are also fine with that.

Comment (nmnobre) [reply makefile]: > We can also just squash-merge the PR.

Indeed, I'm happy with that too.

> If, however, you prefer to have a sequence of commits from the PR

It just seemed like that's what you usually do judging from recent commit history, that's all.
I'll leave it to @tzanio to decide. I'm happy either way really.

**Review by v-dobrev — APPROVED** [review]

File: makefile (hunk 1 above)
Comment (tzanio) [reply makefile]: > I'd like to squash some commits and force push.

Please go ahead and do that

**Review by justinlaughlin — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...
